### PR TITLE
[Gutenberg] Add blog ID to editor session events

### PR DESCRIPTION
**Pull request by hypest**
Following suit with https://github.com/wordpress-mobile/WordPress-iOS/pull/17085, this PR adds the `blog_id` to the editor session events.

This is done essentially by reusing the [AnalyticsUtils.trackWithSiteDetails()](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsUtils.java#L243) helper method.

All editor session events have been augmented with this PR:
* `wpandroid_editor_session_start`
* `wpandroid_editor_session_end`
* `wpandroid_editor_session_switch_editor`
* `wpandroid_editor_session_template_apply`

Note: I had to refactor some of the Analytics code to allow for passing around `AnalyticsTrackerWrapper` to make the methods testable.

### To test

1. Open the block editor.
1. Tap the three-dot menu in the top right.
1. Tap Switch to HTML Mode.
1. Tap the three-dot menu in the top right.
1. Tap Switch to Visual Mode.
1. Close the editor.
1. Verify the following Tracks events include the correct `blog_id`
   property
    * `wpandroid_editor_session_start`
    * `wpandroid_editor_session_switch_editor`
    * `wpandroid_editor_session_end`

### To test the automated tests
1. Comment out the `blog_id` property insertion [at the "source"](https://github.com/wordpress-mobile/WordPress-Android/blob/96b3eecd3145f9f5463620406507c2ef82d1e620/WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsUtils.java#L262)
2. Run the `PostEditorAnalyticsSessionTest` via AndroidStudio or in the CLI via `./gradlew :WordPress:cleanTestWordpressWasabiDebugUnitTest :WordPress:testWordpressWasabiDebugUnitTest --tests "org.wordpress.android.ui.posts.editor.PostEditorAnalyticsSessionTest"`
3. Notice the tests failing with `Argument(s) are different! Wanted:` error where the blog_id is entirely missing from the `Actual invocations have different arguments:` parts.

Another way to test is to:

1. Modify `MapHasEntry("blog_id" to blog_id)` to something like `MapHasEntry("blog_id" to 2222L)`
2. Run the tests as before and notice them error with `Argument(s) are different!` where the actual invocations include `1234L` but the test waiting for `2222L`.
3. 

## Regression Notes
1. Potential unintended areas of impact
As I had to refactor the `AnalyticsUtils.trackWithSiteDetails()` method a bit (to make it more testable via `AnalyticsTrackerWrapper`), tracks that use that method may be impacted if buggy.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

I tried sharing a photo to WP starting from the camera app, and that triggered the `wpandroid_share_to_wp_succeeded` event, which I verified appearing OK in the backend.

3. What automated tests I added (or what prevented me from doing so)
Added tests that verify that the tracks have the `blog_id` property.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt [hunk 1]
```diff
@@ -60,5 +60,13 @@ class AnalyticsTrackerWrapper
         AnalyticsUtils.trackWithSiteDetails(stat, siteModel)
     }
 
+    fun trackWithSiteDetails(
+        stat: Stat,
+        siteModel: SiteModel?,
+        properties: Map<String, Any>
```
Comment (antonis) [inline WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt]: This is more of a suggestion than a change request: What if we added the `properties` parameter with a default value `null` in the function above?

File: WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt [hunk 2]
```diff
@@ -60,5 +60,13 @@ class AnalyticsTrackerWrapper
...
+    ) {
+        AnalyticsUtils.trackWithSiteDetails(stat, siteModel, properties)
```
Comment (antonis) [inline WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt]: We can pass `this` as a 1st parameter to [avoid creating a new instance of the wrapper](https://github.com/wordpress-mobile/WordPress-Android/pull/15465/files#diff-09531fb734dc98890b1902c4bc1ac651ff4a602c4de00a079fe08752a0b5aa2bR284).

**Review by antonis — APPROVED** [review]
Great work @hypest 👍 
I tested the PR on a Pixel 5 (Android 12) and everything worked as expected. The code also looks good 🎉 
ps. I've added a couple of minor comments/suggestions but nothing blocking.

File: WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt (hunk 1 above)
Comment (hypest) [reply WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt]: Oh, your comment made me realize that this (Stat, SiteModel, properties) version of the function is not used anywhere and it's probably code from an intermediate state of my effort here. I shouldn't have committed it. WDYT about removing it altogether?

While at it, it looks like at this stage, [this version of `trackWithEvents`](https://github.com/wordpress-mobile/WordPress-Android/blob/94ceccabb0fa240a87518301d8a002f30f44185f/WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt#L56-L61) has the exact same signature and implementation as [this `track` version](https://github.com/wordpress-mobile/WordPress-Android/blob/94ceccabb0fa240a87518301d8a002f30f44185f/WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt#L41-L43) anyway. WDYT about letting `trackWithEvent` go in favor or a `track` with an optional `properties` parameter? It would look like this: https://github.com/wordpress-mobile/WordPress-Android/compare/gb/add-blog-id-to-editor-session-events...gb/add-blog-id-to-editor-session-events-consolidated-functions

File: WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt (hunk 2 above)
Comment (hypest) [reply WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt]: Hmm, not sure I follow there Antonis 🤔. Can you rephrase or elaborate? Thanks!

File: WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt (hunk 1 above)
Comment (antonis) [reply WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt]: > WDYT about letting trackWithEvent go in favor or a track with an optional properties parameter?

Thats exactly what I was thinking of but I hadn't noticed the method with the exact same signature 👍

File: WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt (hunk 2 above)
Comment (antonis) [reply WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt]: Sorry for the cryptic comment 😄 
I'll try to elaborate. If I'm not missing something we can change the following call
`AnalyticsUtils.trackWithSiteDetails(stat, siteModel, properties)` to
`AnalyticsUtils.trackWithSiteDetails(this, stat, siteModel, properties)` and directly call the [modified trackWithSiteDetails method](https://github.com/wordpress-mobile/WordPress-Android/pull/15465/files#diff-09531fb734dc98890b1902c4bc1ac651ff4a602c4de00a079fe08752a0b5aa2bR243) without creating a [new instance of the `AnalyticsTrackerWrapper `](https://github.com/wordpress-mobile/WordPress-Android/pull/15465/files#diff-09531fb734dc98890b1902c4bc1ac651ff4a602c4de00a079fe08752a0b5aa2bR284)

File: WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt (hunk 1 above)
Comment (hypest) [reply WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt]: Ah, perfect, pushed 8613b41 with the changes.

File: WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt (hunk 2 above)
Comment (hypest) [reply WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt]: Ooooh, I see now, apologies. I was getting caught up in the fact that the [simpler variant](https://github.com/wordpress-mobile/WordPress-Android/blob/96b3eecd3145f9f5463620406507c2ef82d1e620/WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsUtils.java#L281-L285) is used elsewhere anyway and we wouldn't be able to remove the in-place instantiation of AnalyticsTrackerWrapper, but now I understand better what you mean:
the call to AnalyticsUtils.trackWithSideDetails from inside the AnalyticsTrackerWrapper doesn't (or better, shouldn't) lead to creating a new AnalyticsTrackerWrapper.

Made the change with 25e4dd4.

**Comment by hypest** [comment]
Thanks for the review @antonis ! I've addressed the comments so far, let me know if you have more feedback, thanks!

File: WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt (hunk 1 above)
Comment (antonis) [reply WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt]: Thank you for the change Stefanos 👍

File: WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt (hunk 2 above)
Comment (antonis) [reply WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt]: Thank you Stefanos 🙇

File: WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt [hunk 3]
```diff
@@ -38,8 +38,13 @@ class AnalyticsTrackerWrapper
         AnalyticsTracker.track(stat, properties + feature.toParams())
     }
 
-    fun track(stat: Stat, site: SiteModel?) {
-        AnalyticsUtils.trackWithSiteDetails(stat, site)
+    @JvmOverloads
+    fun track(stat: Stat, site: SiteModel?, properties: Map<String, Any?>? = null) {
+        if (properties == null) {
+            AnalyticsUtils.trackWithSiteDetails(stat, site)
```
Comment (antonis) [inline WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt]: Continuing our [discussion here](https://github.com/wordpress-mobile/WordPress-Android/pull/15465#discussion_r741257414) wdyt of removing the `if` clause completely and use in all cases the call:
`AnalyticsUtils.trackWithSiteDetails(this, stat, site, properties)`?
The [`AnalyticsUtils.trackWithSiteDetails(stat, site)`](https://github.com/wordpress-mobile/WordPress-Android/blob/25e4dd4cb2250ee9a9dd06e9069a558a2258abb8/WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsUtils.java#L232) calls:
[`trackWithSiteDetails(stat, site, null)`](https://github.com/wordpress-mobile/WordPress-Android/blob/25e4dd4cb2250ee9a9dd06e9069a558a2258abb8/WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsUtils.java#L281) and instantiates an `AnalyticsTrackerWrapper`.

Comment (hypest) [reply WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt]: Good point and makes sense! Besides, `trackWithSiteDetails(..., properties)` checks for null properties anyway so, it covers the conditional 👍. Done with 5c34afa.

Comment (antonis) [reply WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt]: Thank you for all the changes Stefanos 🙇 
I don't have any further comments on this PR 🎉

Comment (hypest) [reply WordPress/src/main/java/org/wordpress/android/util/analytics/AnalyticsTrackerWrapper.kt]: Grand! Thanks for the reviews and keen eye Antonis! I will assign the PR to myself and go ahead with the merge 👍
